### DISCUSSION TITLE
require() confusion

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#5463] [Q&A] [answered]
I'm looking for the best/easiest way to bootstrap my library. Here's what I have:

```js
// pb_hooks/index.pb.js

onAfterBootstrap((e) => {
  return require(`pocketpages`).AfterBootstrapHandler(e)
})

routerUse((next) => {
  return require(`pocketpages`).MiddlewareHandler(e)
})
```

I'm a little confused about the function nesting in this example:

```js
routerUse((next) => {
    return (c) => {
        console.log(c.path())
        return next(c)
    }
})
```

This is different than expressjs, where `next` is passed along with the context:

```js
app.use((req, res, next) => {
  console.log('Time:', Date.now())
  next()
})
```
If every middleware requires returning a function, why not just collapse the inner function?

I'm also not sure if there is a way to run any middleware initialization code just once instead of with every request. I'd like to do something like:

```js
function MakeMyMiddlewareHandler = ()=>{
   // init code here - require()'s, etc
   return next => {
      return e=> {
         ...handle the request
      }
   }
}
routerUse(MakeMyMiddlewareHandler())
```

But that doesn't seem to work because the outer scoped variables aren't available inside the request handler when it's called.

> Each handler function (hook, route, middleware, etc.) is serialized and executed in its own isolated context as a separate "program". 

Is that why? I'd like to understand if there is a better way to accomplish any of this so my init code doesn't need to run every time. Right now, I'm using `$app.cache` but that can only hold objects, not code.

**Comment by ganigeorgiev** [comment]
The `routerUse` argument follows the same format as the regular `echo.Use`, aka. it expects [`echo.MiddlewareFunc`](https://pkg.go.dev/github.com/labstack/echo/v5#MiddlewareFunc).

We use the same signature as in Go for consistency and so that we can also pass the PocketBase Go builtin middlewares alongside the user defined one without having to create custom dedicated middlewares just for the goja integration (_which will also require extra documentation; for the majority of the bindings we rely on the existing Go docs by extracting them with [tygoja](https://github.com/pocketbase/tygoja)_).

The goal of the goja integration was to be as close as possible to the Go APIs to make the eventual transition to Go easier, ex. changing only the casing of the accessed fields, adding explicit error handling and other minor things without heavy adjustments on the actual application business logic.

> I'd like to understand if there is a better way to accomplish any of this so my init code doesn't need to run every time.

The shown example is too generic and I'm not sure that I understand what actually is the desired behavior. The code in your last example should be fine if the returned `echo.MiddlewareFunc` function doesn't use variables defined before that (aka. the cited scope limitation). This is because when you call `MakeMyMiddlewareHandler()` the returned function is serialized, aka. converted to plain string, and that string is later executed separately (_without any knowledge about its previous parent scope_), as part of a wrapped Go middleware so that it can safely run within the request's goroutine (_it is a little outdated but you can find an overview how the goja integration works in https://github.com/pocketbase/pocketbase/discussions/2825#discussioncomment-6376210; the actual middlewares wrapping code could be found in [`plugins/jsvm/binds.go`](https://github.com/pocketbase/pocketbase/blob/5547c0deded8f9cc329cd6f0670aef19e2a3001a/plugins/jsvm/binds.go#L243-L284)_).

Side-notes: 
- `$app.cache()` (and its Go equivalent `app.Cache()`) is deprecated and aliased to `app.store()`. The deprecated method is removed with the refactoring and will no longer be available with PocketBase v0.23.0.
- echo is replaced with a thin wrapper around the std Go 1.22 mux enhancements. The route handlers, middlewares and hooks now all share similar signature, aka. `func(e) { e.Next() }` (_there is also optional priority support_).

**Reply by benallfree** [reply]
Ok so for example,

```
pb_hooks/
   index.pb.js
node_modules/
   pocketpages/
      index.js
      debug.js
```

```
// node_modules/pocketpages/index.js

const MiddlewareHandler = next => {
   const { dbg } = require('./debug')
   return c => {
      return next(c)
   }
}
```

```js
// index.pb.js

routerAdd( require('pocketpages').MiddlewareHandler ) // Does not work ('debug' module not found)

routerAdd( next => {
   return require('pocketpages').MiddlewareHandler(next) // Works (finds 'debug' module)
})
```

Can you provide some insight?

**Reply by benallfree** [reply]
Also, is there a way to debug module path loading? Something like [NODE_DEBUG=module](https://nodejs.org/api/cli.html#node_debugmodule)

It would be super helpful to have more visibility into why modules fail to load.

**Reply by benallfree** [reply]
Sorry one more thing - a way to specify additional module search paths would be hugely helpful.

`pocketbase serve --jsModuleDir=.,path1,path2,path3`

**Reply by ganigeorgiev** [reply, marked as answer]
I'm sorry, I still don't understand what you are trying to do, nor where the confusion comes from.

`routerAdd` is for registering routes. `routerUse`/`routerPre` is for registering middlewares. Both functions expects different arguments.

Again, the middlewares receive one routing handler and expects to return a new one, aka. to be of [`echo.MiddlewareFunc` type](https://pkg.go.dev/github.com/labstack/echo/v5#MiddlewareFunc).

I don't understand your request but there are no plans for now to change or fork the current default goja modules loading behavior. I don't want to create or maintain a JS runtime. The JSVM is just a plugin/addon for PocketBase and I don't want to depend too much on it.
